[IMP] core: Environment._ for translations and LazyTranslate

kmagusiak opened the pull request:
Python translations need to know the language and the module to find the translated term. Today's `_` function must resolve both of these values from the current stack by using some heuristics.
These heuristics don't always work properly: to find the language we look in the caller's scope to find it in the context or to find the environment from which we can read lang.
For the module, if the function is used outside of an addon, we default to 'base'.

We add a function `Environment._` for translations. That function uses its `lang` property for the language without resorting to inspecting the call stack. You can also pass the module, but we can still resolve it dynamically.

For lazy translations, we create a new `LazyTranslate` factory that you can use as `_lt = LazyTranslate(__name__)` in any addon. This will resolve the addon name once per file from the name of the package without going through `inspect`. You can then use `Environment._` to get the translation or `str()` to resolve the lanugage dynamically.


odoo/enterprise#67528
odoo/documentation#10421
task-4034999

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/174844.png)](https://mergebot.odoo.com/odoo/odoo/pull/174844)

kmagusiak commented:
# Performance improvement

## Env translation
```python
def testme(self, f):
    return f("some string")

%timeit testme(self, _)
%timeit testme(self, self.env._)
TODO

```

## Lazy
```python
# before
%timeit _lt("hello world")
# 45.3 μs ± 201 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# after
%timeit lt = LazyTranslate('odoo.addons.test')
# 2.63 μs ± 27.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
%timeit lt("hello world")
# 4.05 μs ± 50 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
For 10 translations we go from 450μs to 43μs: 10x improvement.

tde-banana-odoo commented:
``Envrionment`` ?

kmagusiak commented:
> `Envrionment` ?

Class `odoo.api.Environment` the instance is available on models as `self.env`

tde-banana-odoo commented:
No, it's just that there is a typo in commit header message ^^

oomsveta reviewed (commented):
I think the changes might break the current Pylint checks.

https://github.com/odoo/odoo/blob/2ee1603ab4b9c7f7ad3fb5f77dde77618ddd1e89/odoo/addons/test_lint/tests/_odoo_checker_gettext.py#L56

I believe that `isinstance(node.func, astroid.Name)` ensures that the function is a simple, unqualified identifier. Meaning it matches `_()`, but not `self.env._()`. Could you add some illegal code on purpose and see if it's still detected by the linter?

kmagusiak commented:
> I think the changes might break the current Pylint checks.
> 
> https://github.com/odoo/odoo/blob/2ee1603ab4b9c7f7ad3fb5f77dde77618ddd1e89/odoo/addons/test_lint/tests/_odoo_checker_gettext.py#L56
> 
> I believe that `isinstance(node.func, astroid.Name)` ensures that the function is a simple, unqualified identifier. Meaning it matches `_()`, but not `self.env._()`. Could you add some illegal code on purpose and see if it's still detected by the linter?

Indeed, I fixed the linter and added a test to check if both functions are detected in the same way.

rco-odoo commented:
The performance improvement of `_lt("hello world")` does not make much sense.  If you use `_lt` is in the module's toplevel code, it is run once and only for the module part.  The real translation occurs when you call `str(msg)`, which will introspect the stack frame to retrieve the target language.  So overall the actual translation speedup is not that big, is it?

rco-odoo commented on odoo/tools/image.py:
```diff
@@ -85,7 +85,7 @@ def __init__(self, source, verify_resolution=True):
                 self.image = Image.open(io.BytesIO(source))
             except (OSError, binascii.Error):
-                raise UserError(_("This file could not be decoded as an image file."))
+                raise UserError(_lt("This file could not be decoded as an image file."))
```
This is not correct.  First, `_lt()` does not return a `str` but an object that returns the translated string.  And second, the string conversion that _actually_ performs the translation will be done somewhere in another stack frame that _possibly_ doesn't have a `self.env`, or that has a `self.env` _with another language in context_.

kmagusiak replied:
Indeed it is not correct. Actually, `odoo.tools` should not try to translate anything except if it has explicitly some record or env. There are other usages like this and I just aligned this one: `_` is obviously incorrect, but `_lt` may resolve to something and if it does not find a language, it will log a warning.
Ideally, tools should raise some custom exceptions which could be caught and re-raised as translated exceptions in business code; otherwise `env` could be passed always as argument.

Do you prefer?
1. I leave as-is, until it is fixed if needed.
2. I add env to nearly all functions using `_lt` in tools.

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -443,155 +444,216 @@ def translate_sql_constraint(cr, key, lang):
+        module, lang = self._get_translation_source(1)
+        return get_translation(module, lang, source, args or kwargs)
+
+    def _get_translation_source(self, stack_level: int, module: str = '', lang: str = '', default_lang: str = '') -> tuple[str, str]:
```
I have the feeling that this method and other associated ones could be static functions used by both `_` and `_lt`.

kmagusiak replied:
`GettextAlias` has no state, so even the class could be removed. I think it was like this to avoid adding more functions to the main namespace. I will simplify it.

rco-odoo commented on odoo/tools/translate.py:
```diff
@@ -443,155 +444,216 @@ def translate_sql_constraint(cr, key, lang):
+    def _get_translation_source(self, stack_level: int, module: str = '', lang: str = '', default_lang: str = '') -> tuple[str, str]:
+        if not (module and lang):
+            frame = inspect.currentframe()
+            while stack_level >= 0:
+                stack_level -= 1
+                frame = frame.f_back
```
```suggestion
            for _index in range(stack_level + 1):
                frame = frame.f_back
```

rco-odoo commented on odoo/api.py:
```diff
@@ -751,7 +751,7 @@ def lang(self):
         lang = self.context.get('lang')
         if lang and lang != 'en_US' and not self['res.lang']._get_data(code=lang):
-            raise UserError(_('Invalid language code: %s', lang))
+            raise UserError(self(context={'lang': 'en_US'})._('Invalid language code: %s', lang, _mod=__name__))
```
Setting the language to `en_US` prevents translation.  So it should be equivalent to this, shouldn't it?
```suggestion
            raise UserError('Invalid language code: %s' % lang)
```

MiquelRForgeFlow replied:
```suggestion
            raise UserError(_('Invalid language code: %s'), lang)
```

kmagusiak replied:
I will go with a simple f-string and disable the lint warning with explanation why.
Using `_` if you don't have a language will just fail.

rco-odoo commented on odoo/api.py:
```diff
@@ -766,6 +766,45 @@ def _lang(self):
             lang = '_' + lang
         return lang
 
+    def _(self, source: str | LazyGettext, *args, _mod='', **kwargs) -> str:
```
This parameter name `_mod` sucks, use `_module` instead.
Also, I don't see the point of introducing it if you don't use it.  The only place you used it was useless.
And I am not even sure it is good practice: moving code from one module to another may lead to translation inconsistencies, if you don't think about checking those `_mod=...`.

kmagusiak replied:
Removed, makes no sense for now.
We talked with ryv about whether transaction should depend on the module, just like what is done in JS. Still, there are context-dependent translations that we would need to handle with some parameter: today it is the module (which we also use to load only the used po files).

rco-odoo commented on odoo/api.py:
```diff
@@ -699,7 +699,7 @@ def company(self):
                 user_company_ids = self.user._get_company_ids()
                 if set(company_ids) - set(user_company_ids):
-                    raise AccessError(_("Access to unauthorized or invalid companies."))
+                    raise AccessError(self._("Access to unauthorized or invalid companies."))
```
Yes, this one was wrong and is now correct :+1:

rco-odoo commented on odoo/api.py:
```diff
@@ -766,6 +766,45 @@ def _lang(self):
+        :return: The transalted string
+        """
+        module = _mod
+        lang = self.lang or self.user.lang or 'en_US'
```
Not sure falling back on `self.user.lang` is a good idea from a performance point of view.  Moreover, we don't use that fallback for data translations, why would we use it for code translations?  IMO for translations to be effective, you must set up the context properly.

kmagusiak replied:
For consistency, I will remove it.
It was the equivalent of "Last resort: attempt to guess the language of the user".

rco-odoo commented on odoo/fields.py:
```diff
@@ -3034,7 +3034,7 @@ def _description_domain(self, env):
                 field_to_check = 'company_ids'
             else:
-                _logger.warning(_(
+                _logger.warning(self.env._(
```
`self` is a field...
```suggestion
                _logger.warning(env._(
```

rco-odoo reviewed (commented):
I am quite skeptical about the claimed performance improvement, since you actually don't take really advantage of it.  The only case where it is effective is when no translation is needed (`lang` is empty or `lang="en_US"`).  It's only a shortcut for cases that are not common on a non-English database.

Also, using `_lt()` for exception messages feels like really bad practice to me, since the translation does not occur where it used to do.  Relying on some lazy evaluation that depends on the execution environment is a recipe for bugs IMO.

kmagusiak commented:
> I am quite skeptical about the claimed performance improvement, since you actually don't take really advantage of it. The only case where it is effective is when no translation is needed (`lang` is empty or `lang="en_US"`). It's only a shortcut for cases that are not common on a non-English database.

Performance improvement is minor (if any), but the goal was to show it is not worse.
Note that I see most of people working in the backoffice using en_US :)

> Also, using `_lt()` for exception messages feels like really bad practice to me, since the translation does not occur where it used to do. Relying on some lazy evaluation that depends on the execution environment is a recipe for bugs IMO.

I feel like lazy translations are misunderstood: they exist so that we can define the translation only once and use it in multiple places in the same file. The right thing to do is to run `env._(some_lazy_translation)` instead of returning them from any module. Using them inside exceptions is also a bad practice for the reasons you give, but I felt it was out of scope of this PR.
Actually `str(lazy_translation)` should log a warning in the future or we could remove lazy translations altogether (if needed, make a helper function that returns a translated string).

rco-odoo commented on addons/base_iban/models/res_partner_bank.py:
```diff
@@ -28,21 +31,21 @@ def get_bban_from_iban(iban):
     iban = normalize_iban(iban)
     if not iban:
-        raise ValidationError(_("There is no IBAN code."))
+        raise ValidationError(_lt("There is no IBAN code."))
```
Wouldn't it be better to add `env` to the API of `validate_iban`?

rco-odoo commented on addons/mail/tools/parser.py:
```diff
@@ -3,11 +3,11 @@
 
 
-def parse_res_ids(res_ids):
+def parse_res_ids(res_ids, env):
```
That's the right way :+1:

rco-odoo commented on addons/phone_validation/tools/phone_validation.py:
```diff
@@ -18,15 +19,15 @@ def phone_parse(number, country_code):
         except phonenumbers.phonenumberutil.NumberParseException as e:
             raise UserError(
-                _('Unable to parse %(phone)s: %(error)s', phone=number, error=str(e))
+                _lt('Unable to parse %(phone)s: %(error)s', phone=number, error=str(e))
```
Try to pass some `env` in parameters, if possible.

kmagusiak replied:
Too many usages.

rco-odoo commented on addons/project_timesheet_holidays/__init__.py:
```diff
@@ -20,23 +19,23 @@ def post_init(env):
             if not internal_projects_by_company_dict:
                 internal_projects_by_company_read = project.search_read([
-                    ('name', '=', _('Internal')),
+                    ('name', '=', project.env._('Internal')),
```
You have `env` as a parameter, you don't need to use `project`.
```suggestion
                    ('name', '=', env._('Internal')),
```

rco-odoo commented on addons/stock/models/stock_rule.py:
```diff
@@ -383,6 +383,7 @@ def _get_lead_days(self, product, **values):
         :return: the cumulative delay and cumulative delay's description
         :rtype: tuple[defaultdict(float), list[str, str]]
         """
+        env = self.env
```
You could define `_` instead
```suggestion
        _ = self.env._
```

rco-odoo commented on odoo/service/model.py:
```diff
@@ -74,9 +75,9 @@ def execute(db, uid, obj, method, *args, **kw):
-    unknown = _('Unknown')
-    model = DotDict({'_name': unknown.lower(), '_description': unknown})
-    field = DotDict({'name': unknown.lower(), 'string': unknown})
+    unknown = _lt('Unknown')
```
Why not `env._`?

kmagusiak commented on odoo/tools/image.py:
```diff
@@ -422,7 +422,7 @@ def binary_to_image(source):
         return Image.open(io.BytesIO(source))
     except (OSError, binascii.Error):
-        raise UserError(_("This file could not be decoded as an image file."))
+        raise UserError(_lt("This file could not be decoded as an image file."))
```
Many usages

rco-odoo reviewed (approved):
Looks fine now.

rco-odoo commented:
@robodoo r+

robodoo commented:
@kmagusiak @rco-odoo unable to stage: merge conflict

rco-odoo commented:
@robodoo delegate+

kmagusiak commented:
@robodoo r+

moylop260 commented:
Hi, the following lint detects all the old cases using `_` method:
 - https://github.com/odoo/odoo/pull/193345

And there are a lot of occurrences:
 - https://gist.github.com/moylop260/71311407f4a00f9a942a4e51757b00f6

My question is, what is the plan with all these sentences using the old way `_`?

I mean, it will be migrated to new `self.env._` or `_lt`?
